Input Rewrite

TimonPost opened the pull request:
**Note, this design is not accurate anymore. Please have a look further into the discussion.**

## New
- Introduced the things discussed [here](https://github.com/crossterm-rs/crossterm/issues/257#issuecomment-539403231). It already addresses some of your comments. There is still some work left for this to be finished. I would like to hear your ideas on how to future envolve this?
- Removes `Terminalnput` because it would become boilerplate code.
- Introduces: ~`EventChannel`~, ~`EventConsumer`~, `EventPool`, ~`EventIterator`~, `EventSource`.
- ~Allows iterating specific events.~
- Allows swapping out event source (testing)
- ~Introduces a single producer multiple consumer idea for input sharing~

_usage example_
```rust
use crossterm::{event_stream, poll_event, RawScreen};

fn main () {
    let r = RawScreen::into_raw_mode().unwrap();

    let mut stream = event_stream();

    while true {
        poll_event();

        for event in stream.key_events() {
            println!("{:?}", event);
        }
    }
}
```

~## Area of Discussion:~

~- Should we store local cache in `EventStream`? The idea behind this is that when reading from the channel we drain the events at the same time? When we read we can store them locally so that we can consume them later on and no events will be lost when for example the user uses `InputStream::key_events` first and then `InputStream::events()`.~
~- Should the poll function only read a single event or read everything available in the buffers.~

~## Todo~

- [ ] implement mouse event enable/disable
- [ ] implement read line
- [ ] More tests
- [ ] Documentation

~## Diagram:~

![image](https://user-images.githubusercontent.com/19969910/67392214-6df73200-f5a0-11e9-9949-ff27cb8d8aaa.png)


## Definitions

### SPMC
~I use shrev `EventChannel` for this, this channel can have multiple readers and a single writer. This is exactly what we want. I tried with normal MPMC channels before, but this didn't work out well. It required a lot of manual thread-safe work like putting everything in arc, mutexes, etc.~


### Event Pool

**EventPool** 

- It is stored statically in with `lazy static`, there should only be one instance of this pool. 
~- It is the owner of the SPMC channel.~
- It is the only one allowed to grant read/write access to the owned event channel.
~- It can create multiple consumers and a single producer. ~
- Access to `EventPool` is secured with `RwLock`. Therefore consumers can be created while a write lock is obtained. 
~- Behaves similar to `Stdin`~
- Only reads a single input event on `poll`. This allows the user to create its own async wrapper. The 

~**EventPoolReadLock**~

~- Provides read access to the `EventPool`~

**EventPoolWriteLock**

- Provides write access to `EventPool`

~**event_stream**~

~-A static function that returns an event stream that can be used to read input events with.~

**poll_event**

~A static function that polls for input from the underlying input source.
An input event will be replicated to all consumers aka streams if an input event has occurred.
This poll function will block read for a single keypress.~

### Input Source
`InputPool` has one input source. This input source can be anything. In the tests I use the `FakeInputSource`, this allows us to fake input data, this might be useful for the user as well for testing it's application. 

- An input source does a block read for one key event. 
- Input source is called by `InputPool::poll()`

The PR contains three input sources: `TtyEventSource`, `WinApiEventSource`, `FakeEventSource`

~### Event Stream~

~- Has it's own SPMC consumer, can't write, only read.~
~- It can iterate specifically over certain events, like a mouse, keyboard. This is a great feature I like to support. Because most often people only want to read for up, down, left, right arrows for example.~

prabirshrestha commented:
Are there plans to also support resize events so that we can re-draw when the window size changes.

TimonPost commented:
Yes the plans are there, we are going to see if we can get something to work with mio polling.

zrzka commented:
I'm not sure if this is the best approach we can get. I mean, implementing something in a PR and discussing it. I'd rather see a list of written requirements:

* How the user will use it with _real_ examples (fake types),
* How the user can achieve sync/async. Do we even want to provide both ways? If not, should we provide async only?
* How the user can wrap it to have `futures::Stream`.
* Do we want to allow users to create multiple event streams? If yes, what's the use case and reasoning behind yes?
* How the user can simulate `poll` aka wait for an event with/without timeout.
* Do we want to catch signals and send them via `Event` too.
* And probably lot of other questions.

Once we will have answers here, we can think about how to implement it with our own internal requirements like:

* How can the `cursor::position` fn retrieve `InternalEvent::CursorPosition`.
* How can we do text area resize event.
* Do we really need the reading thread? Can't we achieve it with polling and expose it to the user somehow?

---

My comments on this PR (not exhaustive) ...

### General

* `event_pool`, `event_source`, ... -> `event::pool`, `event::source`, ...

### `input::event_source::fake::FakeEventSource`

* Should be gated for tests only and moved to actual `tests` module

### `input::event_source::tty::TTYEventSource`

* `TTYEventSource` -> `TtyEventSource`
* `TTYEventSource::new()` shouldn't `unwrap` to avoid panics
* Should produce `InternalEvent`, not `InputEvent`

### `input::event_stream::EventStream`

* What's the purpose of the `event_cache` field?
* Aren't `key_events` & `events` mutually exclusive? Seems that when I'll have two iterators, one for `key_events` and another one for `events`, `key_events` one will drain all the key events and `events` iterator wont produce them?
* What's the purpose of `key_events` & `mouse_events`? Are we expecting that user is going to use multiple event consumers/iterators? Why not just one stream of all `events`?
* Generally speaking, why the user can obtain multiple event streams? IMHO user should be able to get just one. We have talked about single producer & multiple consumers, but I think it should be internal not a public feature. For example for the cursor position - internally, we can get another stream and wait for `CursorPosition`. But why the crate user is allowed to do it?
* How can I obtain `InternalEventStream`? For cursor position for example.
* `EventStream` should be a wrapper of `InternalEventStream`, which filters special internal events out only.

### `input::events::`

* I'd rename `InputEvent` to `Event`

#### `KeyEvent`

Something like:

```rust
struct KeyEvent {
  code: KeyCode,
  modifiers: KeyModifiers
}

enum KeyCode { Backspace, Enter, ... }

bitflags! {
    struct KeyModifiers: u8 {
        const Shift = 0b00000001;
        const Control = 0b00000010;
        ...
        const ShiftControl = Self::Shift.bits | Self::Control.bits;
        ...
    }
}
```

### `input::event_pool::EventPoll`

* What's the purpose of the `poll` method? Why the user is forced to call it? I mean, I understand why you did it in this way, but user shouldn't be forced to call it. Iterator can return `None`.
* I mean, it would be useful if user can `poll` with some kind of timeout. Something like:
  * `poll` without timeout - sits there till event is produced and then it will be guaranteed that the iterator will return an event, user can even unwrap,
  * `poll` with timeout - sits there till event is produced / timeout elapsed, nothing is guaranteed, iterator can return `None`.
* Is there any problem with calling `enable_mouse_events` automatically / internally? Does it brake anything, ...?

TimonPost commented:
> What's the purpose of the `event_cache` field

When only reading events like `KeyEvents` from a channel or in this case an `EventChannel` we have to read all events on the channel and only filter on the `KeyEvents`, but what if the user wants to read key events and mouse events after each other? We would have drained them from the `EventChannel` before and those mouse events are all lost. So we can read events from the `EventChannel` and write those to the `event-cache` field which then can be used to itreate over as many times that is needed. Each `EventConsumer` has it's own internal cache and can iterate over that and apply different masks to read certain events.

> Aren't key_events & events mutually exclusive? It seems that when I'll have two iterators, one for key_events and another one for events, key_events one will drain all the key events and events iterator won't produce them?

As I thought of it is that the `EventIterator<InputEvent>` iterates over all events, e.g. MouseEvent, KeyEvent, ResizeEvent etc. The `EventIterator<KeyEvent>` only iterates over the `KeyEvent`.

> What's the purpose of key_events & mouse_events? Are we expecting that user is going to use multiple event consumers/iterators? Why not just one stream of all events?

The purpose of this is that the user can iterate over specific events. While ignoring others. 

> Generally speaking, why the user can obtain multiple event streams? IMHO user should be able to get just one. We have talked about single producer & multiple consumers, but I think it should be internal not a public feature. For example for the cursor position - internally, we can get another stream and wait for CursorPosition. But why the crate user is allowed to do it?

Why wouldn't we support and limit our API to only support one stream while our internals already does support it? 


> How can I obtain InternalEventStream? For cursor position for example.
Not yet supported, currently, it is based on `InputEvent`, basilcy all ocurences of `InputEvent` can be changed to `InternalEvent`. `EventStream` would get a new function `EventStream::input_events()` which returns `EventIterator<InputEvent>` and `EventStream::events()` would return `EventIterator<InternalEvents>`. Also the event source traid should be changed to return `InternalEvent` like  `EventSource::read_event() : InternalEvent`. 

> EventStream should be a wrapper of InternalEventStream, which filters special internal events out only.

See above answer.


> What's the purpose of the poll method? Why the user is forced to call it? I mean, I understand why you did it in this way, but user shouldn't be forced to call it. Iterator can return None.

Good one, I am not sure about this yet. We can implement `poll` inside the `EventIterator` actually.  The purpose of `EventPool::poll` is to give the user the control to polling and therefore give it control on how to poll. If he wants async he can via this method. 

>  It would be useful if user can poll with some kind of timeout. Something like

Yes, this would be a great feature, I left it out because winapi is the problem here, we can implement it but it does not support cancelation of the read quest by default.

> Is there any problem with calling enable_mouse_events automatically / internally? Does it brake anything, ...?

`MouseEvents` give a lot of noise since there are a lot of them when turning on the monitoring. I am not sure if this is a problem, and if we can just enable them always.

TimonPost commented:
Used your api examples, I haven't tested unix completely yet, altough windows seems to work.

```rust
use crossterm::{
    input::{poll, read},
    Event, KeyEvent, RawScreen,
};
use std::time::Duration;

fn main() {
    let r = RawScreen::into_raw_mode().unwrap();

    read_async();
}

fn sync_read1() {
    loop {
        match read() {
            Ok(event) => {
                if should_stop(&event) {
                    break;
                }
                println!("{:?}", event);
            }
            Err(_) => {
                // read() error
            }
        }
    }
}

fn sync_read2() {
    loop {
        match poll(None) {
            Ok(_) => {
                match read() {
                    Ok(event) => {
                        if should_stop(&event) {
                            break;
                        }
                        println!("{:?}", event);
                    }
                    Err(_) => {
                        // read() error
                    }
                }
            }
            Err(_) => {
                // poll() error
            }
        }
    }
}

fn sync_read3() {
    loop {
        match poll(None).and_then(|_| read()) {
            Ok(event) => {
                if should_stop(&event) {
                    break;
                }
                println!("{:?}", event);
            }
            Err(e) => println!("{:?}", e),
        }
    }
}

fn read_async() {
    loop {
        match poll(Some(Duration::from_millis(100))) {
            Ok(true) => {
                // Event available - read() wont block
                match read() {
                    Ok(event) => {
                        if should_stop(&event) {
                            break;
                        }
                        println!("{:?}", event);
                    }
                    Err(_) => {
                        // read() error
                    }
                }
            }
            Ok(false) => {
                // Event not available, but 100ms timeout expired
                // 10fps if there's no event available
            }
            Err(_) => {
                // poll() error
            }
        }
    }
}

fn should_stop(event: &Event) -> bool {
    *event == Event::Keyboard(KeyEvent::Esc)
}
```

TimonPost commented:
New changes are made. And API is slightly modified as discussed in discord. 

# Public Exposed API
We expose three functions in the public API:

## Exposed Functions

```rust
/// Polls during an given duration for ready events.
///
/// This function takes in an optional duration.
/// * `None`: will block indefinitely until an event is read.
/// * `Some(duration)`: will block for the given duration.
///
/// The following value can be returned:
/// * `Ok(true)`: in case an event is ready.
/// * `Ok(false)`: in case the given duration is elapsed.
/// * `Err(err)`: in case of an error.
pub fn poll(timeout: Option<Duration>) -> Result<bool> { }

/// Reads a single event.
///
/// This function will block until an event is received.
/// Use [poll](LINK) for ready events.
pub fn read() -> Result<Event> { }

/// Changes the default `EventSource` to the given `EventSource`.
///
/// This might be usefull for testing purposes.
pub fn swap_event_source(new: Box<dyn EventSource>) { }
```

## Exposed Types
* `EventSource` (implemented on `FakeEventSource`)
* `FakeEventSource` (fake event source, can be used for testing purposes)

# Internals

### EventPoll Trait
An interface for polling event readiness and reading events.

```rust
pub trait EventPoll {
    type Output;

    fn poll(&mut self, timeout: Option<Duration>) -> Result<bool>;

    fn read(&mut self) -> Result<Self::Output>;
}
```
Implemented for structs: `EventReader` and `InternalEventReader`.

### EventSource Trait

 An interface for trying to read an `InternalEvent` within an optional `Duration`.

```rust
pub trait EventSource: Sync + Send {
    /// Tries to read an `InternalEvent` within the given duration.
    ///
    /// This function takes in an optional duration.
    /// * `None`: will block indefinitely until an event is read.
    /// * `Some(duration)`: will block for the given duration.
    ///
    /// Returns:
    /// `Ok((true, Some(event)))`: in case an event is ready.
    /// `Ok((false, None))`: in case an event is not ready.
    fn try_read(
        &mut self,
        timeout: Option<Duration>,
    ) -> crate::Result<(bool, Option<InternalEvent>)>;
}
```

Used by structs that are platform-specific implementations and provide `InternalEvents`. Implemented by `WinApiEventSource`, `TtyInternalEventSource`, `FakeEventSource`.

### EventReader & InternalEventReader
Abstracts away platform-specific polling mechanism. 
Implements the `EventPoll`  trait.

### Event Pool

- It is stored statically in with `lazy static`, there should only be one instance of this pool. 
- Access to `EventPool` is secured with `RwLock`. We could consider `Mutex` because currently there are not mutable requests for this type. 
- Offers an interface to interact with `EventReader` and `InternalEventReader`.

**EventPoolWriteLock**

- Provides write access to `EventPool`

# Demo
see `/examples/input_demo.rs` (will be removed)

- run `cargo run --example input_demo`

zrzka commented:
@TimonPost I started writing comments, but there were lot of them. Decided to do a review in another form. Check [this PR](https://github.com/crossterm-rs/crossterm/pull/312) and all my commits. A bit of everything there. Or just ...

```bash
$ git diff remotes/origin/input_rewrite..remotes/origin/zrzka/input_rewrite_comments
```

It's a **1st round**, to make it compile, remove warnings, pass CI, clippy, ...

TimonPost commented:
1. Why the removal of `rustfmt.toml`? 
   - I don't mind removing it, I added it because I can easily merge imports now by just switching to nightly and run `cargo fmt`. Also, we can create rules for styling here.
2. What are we going to do with `source::fake::FakeEventSource`. 
    - You moved those to the tests of `pool`, but I think we need this fake reader in the tests of `events.rs` as well. 
    - Maybe we can export `FakeEventSource` with the `#[cfg(test)]` so we can keep it in the same folder as the other `EventSources` and it will only be accessible from the tests.
    - Aren't we going to provide this `FakeEventSource` to the user? 'Maybe' it can be useful for a user to simulate key_events. 
3. I don't want to include all Clippy warnings fixes into this PR. We can open another one for this. This PR already did some important changes to other modules, but I like to keep this PR narrowed down to the input module.
4. I didn't want to include the renames of `InternalEvent` into this PR originally, however, doesn't really matter. So will rename those.
5. Ok for the renames file pool/reader/source.

zrzka commented:
> 1. Why the removal of `rustfmt.toml`? 

Warning free. `merge-imports` is `nightly` feature. I'd rather see it the `rustfmt.toml` in `.gitignore` for now if you're using it. The mandatory test is for `stable` toolchain.

> 2. What are we going to do with `source::fake::FakeEventSource`. 
>    - Maybe we can export `FakeEventSource` with the `#[cfg(test)]` so we can keep it in the same folder as the other `EventSources` and it will only be accessible from the tests.

I moved it, because it was used in one place only. If we need it in more places, we can gate it elsewhere with `#[cfg(test)]`. But it shouldn't be part of the public API.

>   - Aren't we going to provide this `FakeEventSource` to the user? 'Maybe' it can be useful for a user to simulate key_events. 

Why? The goal is to provide `poll` & `read`. This is internal implementation detail. Minimal API.

> 3. I don't want to include all Clippy warnings fixes into this PR. We can open another one for this. This PR already did some important changes to other modules, but I like to keep this PR narrowed down to the input module.

No need to include it, I just wanted to remove all the warnings in my tests.

> 4. I didn't want to include the renames of `InternalEvent` into this PR originally, however, it doesn't really matter to so I will rename those in this PR.

This PR is about major overhaul of the `input` module, right? And it includes `Event`, etc. Personally, I'd like to see this PR output as a well polished, tested and working `input` module with everything related included.

> 5. Ok for the renames file pool/reader/source.

Did it to make it shorter. And I hate underscores :)

TimonPost commented:
> Why? The goal is to provide poll & read. This is internal implementation detail. Minimal API.
To allow the user to simulate events. Although, more exposure to any API will restrict us to make changes in the future. Let's keep it out.

zrzka commented:
There're still some changes that were not merged/commented (only `input` related):

```bash
$ git diff input_rewrite..zrzka/input_rewrite_comments -- src/input
```

---

List of uncommented changes:

* `Event::Unknown`, `MouseEvent::Unknown` removed
  * Removed them, because they're of no use. User is forced to handle them without any value given back. It's just a noise.
* Many `use crate:...` replaced with `use super::...`. Rules ...
  * `use crate::...`
    * Use only for types that are public (stable API), like `Result`
    * Use only if accessing other modules (like `utils` from the `input` module)
  * `use super::...`
    * Use for everything else which boils down to everything inside the `input` module
* Removed `EventSource` & friends from the public API
  * Including removal of methods like `swap_event_source`
* Unmerged clippy fixes (`input` module only)
* `PollTimer` -> `PollTimeout`
  * Timer is a bit misleading, it's not a timer, it's a duration value (timeout).
  * `left_over` -> `leftover` -> [it's a real word](https://dictionary.cambridge.org/dictionary/english/leftover)
  * Module name `poll_timer.rs` -> `timeout.rs`
* `fake` module
  * Discussed partially, but it shouldn't be part of the crate itself and it should be gated with `#[cfg(test)]` and placed elsewhere?

---

Broken CI again

This `let mut reader = InternalEventReader::new()` wont work, because it opens `/dev/tty` by default and it can't open it on CI. In other words, you can't use `new()` (to avoid `TtyInternalEventSource::new().expect("Failed to setup the default event reader.")`).

Here's my recommendation:

```rust
impl Default for InternalEventReader {
    fn default() -> Self {
        #[cfg(windows)]
        let source = WinApiEventSource::new();
        #[cfg(unix)]
        let source =
            TtyInternalEventSource::new().expect("Failed to setup the default event reader.");

        InternalEventReader::new(Box::new(source))
    }
}

impl InternalEventReader {
    /// Constructs a new `InternalEventReader`.
    pub(crate) fn new(source: Box<dyn EventSource>) -> InternalEventReader {
        InternalEventReader {
            event_source: source,
            events: VecDeque::new(),
        }
    }
```

And ...

* `EventReader::new` - `internal_poll: InternalEventReader::default()`
* `EventPoll::new` - `internal_event_reader: InternalEventReader::default()`

... in tests (`get_polling_thread`) ...

```rust
let (event_sender, event_receiver) = channel();
let mut reader = InternalEventReader::new(Box::from(FakeEventSource::new(event_receiver)));
```

... in this way there wont be an attempt to instantiate `TtyInternalEventSource`/`WinApiEventSource` in tests and CI will work.

It's hard to do review when I can't compile the code and the CI is constantly failing.

TimonPost commented:
1 Removed `EventPool`

Motivation: 
a. public and crate level functions: `poll`, `read` were just call forwarders to `EventPool`.
b. My code had a deadlock when reading with `EventReader`  because it also acquires the lock to `InternalReader`. 

2 Moved `EventReader` and `InternalEventReader` into there own `lazy_static`

Motivation:
* see 1
* Locks for readers can be acquired separately when they aren't wrapped by a parent type `EventPool`

3. Moved `poll`, `read` to, and added some functions that first had a place in `EventPool` in, `input.rs`

zrzka commented on .github/workflows/crossterm_test.yml:
```diff
@@ -63,7 +63,7 @@ jobs:
       run: cargo test --features screen --no-default-features --lib -- --nocapture --test-threads 1
       continue-on-error: ${{ matrix.can-fail }}
     - name: Test input Feature
```
Test event feature

zrzka commented on CHANGELOG.md:
```diff
@@ -1,3 +1,16 @@
+
+- `input` module
+    - Rename `InputEvent` to `Event`
+    - Rename `Event::KeyBord` to `Event::Key`
```
Typo in `KeyBord`.

zrzka commented on CHANGELOG.md:
```diff
@@ -1,3 +1,16 @@
+    - Rename `InputEvent` to `Event`
+    - Rename `Event::KeyBord` to `Event::Key`
+    - Rename `crossterm::input` to `crossterm::event`
+    - Rename `input` feature flag to `event`
```
Can we emphasize this change more? It's quite important.

zrzka commented on Cargo.toml:
```diff
@@ -34,4 +34,4 @@ mio = { version = "0.6.19", optional = true }
 [lib]
 name = "crossterm"
-path = "src/lib.rs"
+path = "src/lib.rs"
```
New line at the end of file please :)

zrzka commented on rustfmt.toml:
```diff
@@ -0,0 +1 @@
+merge_imports = true
```
Can we pls remove the file? Unstable feature requiring `nightly`.

```sh
zrzka@Roberts-MacBook-Pro crossterm % cargo fmt
Warning: can't set `merge_imports = true`, unstable features are only available in nightly channel.
Warning: can't set `merge_imports = true`, unstable features are only available in nightly channel.
zrzka@Roberts-MacBook-Pro crossterm %
```

zrzka commented on src/cursor/sys/unix.rs:
```diff
@@ -29,15 +33,43 @@ fn read_position() -> Result<(u16, u16)> {
+                match read_internal() {
+                    Ok(InternalEvent::CursorPosition(x, y)) => {
+                        if !temp_buffer.is_empty() {
+                            while let Some(event) = temp_buffer.pop_front() {
```
This `enqueue_internal` is something I don't like much, will think about it.

But there's one thing - this `while` loop & `enqueue_internal` should be called for every return path - `Ok(false)`, `Err(e)`, ... - in other words - in every place where the `read_position_raw` exits.

zrzka commented on src/cursor/sys/unix.rs:
```diff
@@ -29,15 +33,43 @@ fn read_position() -> Result<(u16, u16)> {
+                        // and enqueue the events back when we read the cursor position.
+                        temp_buffer.push_back(event);
+                    }
+                    e => {}
```
`Err(_) => {}`

zrzka commented on src/cursor/sys/unix.rs:
```diff
@@ -29,15 +33,43 @@ fn read_position() -> Result<(u16, u16)> {
+                ))?;
+            }
+            Err(e) => {
+                println!("{:?}", e);
```
Do not `println!` ... `Err(_) => {}`

TimonPost replied:
Sorry. Used it for testing purposes.

zrzka commented:
Re macOS stable vs nightly. Don't look at nightly, it's just informative. See this ...

https://github.com/crossterm-rs/crossterm/blob/268230c5793bd48fdf098c047efb6783aa540c52/.github/workflows/crossterm_test.yml#L25-L26

... and this ...

https://github.com/crossterm-rs/crossterm/blob/268230c5793bd48fdf098c047efb6783aa540c52/.github/workflows/crossterm_test.yml#L50-L52

... my assumption here is that there's a flaw in GH Actions. We do allow nightly to fail, but there was a red cross. It's now marked with the green checkmark, which is probably a GH Actions bug.

--- 

Here's the diff to make it working/passing on macOS.

```patch
diff --git a/examples/input_demo.rs b/examples/input_demo.rs
index 3c6e0d4..2c41b2c 100644
--- a/examples/input_demo.rs
+++ b/examples/input_demo.rs
@@ -1,4 +1,4 @@
-#![allow(dead_code)]
+#![allow(dead_code, unused_imports)]
 
 use crossterm::{
     event::{poll, read, EnableMouseCapture, Event, KeyEvent},
diff --git a/src/cursor/sys/unix.rs b/src/cursor/sys/unix.rs
index 355ecdb..7ac7f46 100644
--- a/src/cursor/sys/unix.rs
+++ b/src/cursor/sys/unix.rs
@@ -58,7 +58,7 @@ fn read_position_raw() -> Result<(u16, u16)> {
                         // and enqueue the events back when we read the cursor position.
                         temp_buffer.push_back(event);
                     }
-                    e => {}
+                    Err(_) => {}
                 };
             }
             Ok(false) => {
```

TimonPost commented:
We could provide a function as an argument in the `EventPoll`  that can be used as a filter to prevent returning false positives. So `EventReader` passes an `Event` filter, `InternalEventReader`  uses this filter, and only returns when this filter is true.

zrzka commented on src/event/sys/windows.rs:
```diff
@@ -0,0 +1,374 @@
+        VK_ESCAPE => Some(KeyCode::Esc),
+        VK_RETURN => Some(KeyCode::Enter),
+        VK_F1..=VK_F24 => Some(KeyCode::F((key_event.virtual_key_code - 111) as u8)),
+        VK_LEFT | VK_UP | VK_RIGHT | VK_DOWN => Some(match key_code {
```
I'm a bit puzzle with this ... What is the reason to have them grouped, use nested `match` with `_` & `unreachable!` instead of simple ...

```
VK_LEFT => {},
VK_RIGHT => {},
VK_UP => {},
VK_DOWN => {},
```

... without `unreachable!`?

TimonPost replied:
Good one will do, has to do with what we first did here, and slowly cleaning up the parsing.

TimonPost commented:
It reached the final endpoint. Thanks @zrzka for contributing to this PR with me on. We made something very awesome. I can't wait for it to be released. I will do the `crossterm_winapi` release soon. So that we can remove that GitHub dependency.
